Source/API/Extras/CBLIncrementalStore.m

```diff
@@ -2057,103 +2090,122 @@ - (void) processCouchbaseLiteChanges: (NSArray*)changes {
 
-                    // Ensure that a fault has been fired:
-                    [moc willAccessValueForKey: nil];
-                    [context refreshObject: moc mergeChanges: YES];
+    for (NSManagedObjectContext *context in sortedContextRootBeforeMain) {
+        [context performBlock:^{
+            INFO(@"start refresh context : %@", context.parentContext ? @"MAIN" : @"ROOT");
 
-                    for (NSString *relationshipName in moc.entity.relationshipsByName) {
-                        NSRelationshipDescription *relationshipDescription =
-                        moc.entity.relationshipsByName[relationshipName];
-                        if (!relationshipDescription.toMany) {
-                            if (![moc hasFaultForRelationshipNamed:relationshipName]) {
-                                NSManagedObject *destinationObject = [moc valueForKey:relationshipName];
-                                if (destinationObject) {
-                                    // Ensure that a fault has been fired:
-                                    [destinationObject willAccessValueForKey: nil];
-                                    [context refreshObject: destinationObject mergeChanges: YES];
-                                }
-                            }
-                        }
-                    }
-                } else {
-                    [updated addObject: moc];
+            NSMutableSet *refreshedIDs = [NSMutableSet set];
+            NSMutableSet *updatedEntities = [NSMutableSet set];
+
+            for (NSManagedObjectID* moID in updatedIDs) {
+                NSManagedObject* mObj = [context objectRegisteredForID: moID];
 
-                    // Ensure that a fault has been fired:
-                    [moc willAccessValueForKey: nil];
-                    [context refreshObject: moc mergeChanges: YES];
+                if (!mObj) {
+                    mObj = [context objectWithID: moID];
                 }
 
-                [updatedEntities addObject: moc.entity.name];
-            }
-            [userInfo setObject: updated forKey: NSUpdatedObjectsKey];
-            if (inserted.count > 0) {
-                [userInfo setObject: inserted forKey: NSInsertedObjectsKey];
+                for (NSString *relName in moID.entity.relationshipsByName) {
+                    NSRelationshipDescription* rel = moID.entity.relationshipsByName[relName];
+                    if (rel.toMany)
+                        continue;
+
+                    NSRelationshipDescription *invRel = rel.inverseRelationship;
+                    if (!invRel)
+                        continue;
+
+                    NSArray *objIDs = [mObj objectIDsForRelationshipNamed:relName];
+
+                    [refreshedIDs addObjectsFromArray:objIDs];
```

Comment:
Actually the duplicate object is the relationship object. I think NSFetchResultController thinks that the refreshed one is the new one.